feat: Add `num_samples` parameter to `date_range` and `datetime_range`

mcrumiller opened the pull request:
* Closes #7525
* Closes #23424
* Closes #25515

This update adds the `num_samples` parameter to `date_range` and `datetime_range`. The input parameters that define the range are now: `start`, `end`, `interval`, and `num_samples`. A range can be generated using any three of these parameters. In order to keep backwards compatibility, and for the most common usage, the two-parameter combination of `start` and `end` works, and uses a default duration of 1 day.

This update also allows for ranges generated using negative intervals, in much the same way that `linear_space` and `int_range` can generate backwards ranges. Note that the `closed` parameter applies these as _directional_ intervals, meaning that a left-closed series always has the form `[s1, s2, ..., s_n)` even if the values are descending.

This had the minor side effect of enabling negative intervals in `time_range` as well, with no additional effort, so I threw in a quick test for that. We can manually disable that by adding in a check if we don't want that in this PR.

#### Examples

```python
import polars as pl
from polars import col
from datetime import date

df = pl.DataFrame({
    "start": [date(2025, 1, 1), date(2025, 1, 2)],
    "end": [date(2025, 1, 10), date(2025, 1, 11)],
    "num_samples": [5, 6],
})

df.select(
    start_end_samples=pl.datetime_range(start=col("start").min(), end=col("end").max(), num_samples=6),
    start_interval_samples=pl.datetime_range(start=col("start").min(), interval="-2d", num_samples=col("num_samples").min() + 1),
    end_interval_samples=pl.datetime_range(end=col("end").max(), interval="1mo", num_samples=col("num_samples").max()),
    start_end_interval=pl.datetime_range(start=col("start").min(), end=col("end").max(), interval="2d"),
)
# shape: (6, 4)
# ┌─────────────────────┬────────────────────────┬──────────────────────┬─────────────────────┐
# │ start_end_samples   ┆ start_interval_samples ┆ end_interval_samples ┆ start_end_interval  │
# │ ---                 ┆ ---                    ┆ ---                  ┆ ---                 │
# │ datetime[μs]        ┆ datetime[μs]           ┆ datetime[μs]         ┆ datetime[μs]        │
# ╞═════════════════════╪════════════════════════╪══════════════════════╪═════════════════════╡
# │ 2025-01-01 00:00:00 ┆ 2025-01-01 00:00:00    ┆ 2024-08-11 00:00:00  ┆ 2025-01-01 00:00:00 │
# │ 2025-01-03 00:00:00 ┆ 2024-12-30 00:00:00    ┆ 2024-09-11 00:00:00  ┆ 2025-01-03 00:00:00 │
# │ 2025-01-05 00:00:00 ┆ 2024-12-28 00:00:00    ┆ 2024-10-11 00:00:00  ┆ 2025-01-05 00:00:00 │
# │ 2025-01-07 00:00:00 ┆ 2024-12-26 00:00:00    ┆ 2024-11-11 00:00:00  ┆ 2025-01-07 00:00:00 │
# │ 2025-01-09 00:00:00 ┆ 2024-12-24 00:00:00    ┆ 2024-12-11 00:00:00  ┆ 2025-01-09 00:00:00 │
# │ 2025-01-11 00:00:00 ┆ 2024-12-22 00:00:00    ┆ 2025-01-11 00:00:00  ┆ 2025-01-11 00:00:00 │
# └─────────────────────┴────────────────────────┴──────────────────────┴─────────────────────┘

pl.Config.set_fmt_table_cell_list_len(999)
pl.Config.set_fmt_str_lengths(999)
pl.Config.set_tbl_width_chars(9999)

df.select(
    start_interval_samples=pl.datetime_ranges(start="start", interval="1d12h", num_samples=3, closed="right"),
    start_end_samples=pl.datetime_ranges(start="start", end="end", num_samples=3, closed="none"),
)
# shape: (2, 2)
# ┌─────────────────────────────────────────────────────────────────┬─────────────────────────────────────────────────────────────────┐
# │ start_interval_samples                                          ┆ start_end_samples                                               │
# │ ---                                                             ┆ ---                                                             │
# │ list[datetime[μs]]                                              ┆ list[datetime[μs]]                                              │
# ╞═════════════════════════════════════════════════════════════════╪═════════════════════════════════════════════════════════════════╡
# │ [2025-01-02 12:00:00, 2025-01-04 00:00:00, 2025-01-05 12:00:00] ┆ [2025-01-03 06:00:00, 2025-01-05 12:00:00, 2025-01-07 18:00:00] │
# │ [2025-01-03 12:00:00, 2025-01-05 00:00:00, 2025-01-06 12:00:00] ┆ [2025-01-04 06:00:00, 2025-01-06 12:00:00, 2025-01-08 18:00:00] │
# └─────────────────────────────────────────────────────────────────┴─────────────────────────────────────────────────────────────────┘
```

mcrumiller commented on py-polars/src/polars/functions/range/datetime_range.py:
```diff
@@ -80,6 +88,9 @@ def datetime_range(
         Interval of the range periods, specified as a Python `timedelta` object
         or using the Polars duration string language (see "Notes" section below).
+    num_samples
+        Number of periods in the date range. This corresponds to the number of points in
```
Good catch, you're right this is definitely incorrect and confusing.

mcrumiller commented on py-polars/src/polars/functions/range/date_range.py:
```diff
@@ -43,26 +46,32 @@ def date_range(
 
+    A date range may be defined by any three of 'start', 'end', 'interval', and
+    'num_samples'. If only two of `start`, `end`, and `num_samples` are provided, the
+    interval defaults to 1 day.
```
Wouldn't this be the case with the normal positive interval situation too? I'm not sure I see why this is true for negative intervals only.

orlp commented:
> I can add an i64 path there, that may be the best option.

That makes sense to me.

mcrumiller commented:
@orlp main updates since you last looked are in [Remove duplicate enum and use linear_space for start/end/num_samples](https://github.com/pola-rs/polars/pull/25329/commits/8a68bb3e86291476fd66abf5568970dafe7efbca). Two main changes:

1. I discovered that we have identical `ClosedWindow` and `ClosedInterval` enumerations in the codebase, so I made one a type alias of the other, which lets us remove an impl. I think conceptually they are still distinct so it's still good to have both. I updated the range functions to use `ClosedInterval` from `ClosedWindow`.

2. I added `new_linear_space_i64` for the use by the temporal types. I updated the existing `linear_space()` to use this for the temporal path since it will probably be a little bit more performant. `date_range` and `datetime_range` now dispatch to this for the `start`/`end`/`num_samples` path option.

orlp commented on py-polars/src/polars/functions/range/datetime_range.py:
This is no longer true, `start` isn't the lower bound, it's the first output value... except if closed is `right` or `none`... It's all getting rather confusing I'm starting to think this needs a bit more time in the design chamber rather than the PR chamber.
